drivers/nec/display/all.cr

```diff
@@ -0,0 +1,410 @@
+# Status information avaliable:
+# -----------------------------
+#
+# (built in)
+# connected
+#
+# (module defined)
+# power
+# warming
+#
+# volume
+# volume_min == 0
+# volume_max
+#
+# brightness
+# brightness_min == 0
+# brightness_max
+#
+# contrast
+# contrast_min = 0
+# contrast_max
+#
+# audio_mute
+#
+# input (video input)
+# audio (audio input)
+
+class Nec::Display::All < PlaceOS::Driver
+  include Interface::Powerable
+  include Interface::AudioMuteable
+
+  enum Inputs
+    Vga          = 1
+    Rgbhv        = 2
+    Dvi          = 3
+    Hdmi_set     = 4
+    Video1       = 5
+    Video2       = 6
+    Svideo       = 7
+    Tuner        = 9
+    Tv           = 10
+    Dvd1         = 12
+    Option       = 13
+    Dvd2         = 14
+    Display_port = 15
+    Hdmi         = 17
+    Hdmi2        = 18
+    Hdmi3        = 130
+    Usb          = 135
+  end
+  include PlaceOS::Driver::Interface::InputSelection(Inputs)
+
+  # Discovery Information
+  tcp_port 7142
+  descriptive_name "NEC LCD Display"
+  generic_name :Display
+
+  default_settings({
+    volume_min: 0,
+    volume_max: 100
+  })
+
+  @volume_min : Int32 = 0
+  @volume_max : Int32 = 100
+
+  # 0x0D (<CR> carriage return \r)
+  DELIMITER = 0x0D_u8
+
+  @target_input : Input? = nil
+  @target_audio : Audio? = nil
+  @input_double_check : PlaceOS::Driver::Proxy::Scheduler? = nil
+
+  def on_load
+    # Communication settings
+    queue.delay = 120.milliseconds
+    transport.tokenizer = Tokenizer.new(Bytes[DELIMITER])
+    on_update
+  end
+
+  def on_update
+    @volume_min = setting(Int32, :volume_min)
+    @volume_max = setting(Int32, :volume_max)
+  end
+
+  def connected
+    schedule.every(50.seconds, true) do
+      do_poll
+    end
+  end
+
+  def disconnected
+    # Disconnected may be called without calling connected
+    # Hence the check if timer is nil here
+    schedule.clear
+  end
+
+  def power(state : Bool)
+    data = "C203D6"
+    current = self[:power]?
+
+    if current
+      data += "0004" # 0004 = Power Off
+      do_send("command", data, name: "power", delay: 10.seconds, timeout: 10.seconds)
+
+      self[:power] = false
+      logger.debug { "-- NEC LCD, requested to power off" }
+    else
+      data += "0001" # 0001 = Power Off
+      do_send("command", data, name: "power", delay: 5.seconds)
+      self[:warming] = true
+      self[:power] = true
+      logger.debug { "-- NEC LCD, requested to power on" }
+
+      power_on_delay
+      mute_status(20)
+      volume_status(20)
+    end
+  end
+
+  def power?(**options)
+    # options[:emit] = block if block_given?
+    do_send("command", "01D6", **options)
+  end
+
+  def switch_to(input : Input)
+    @target_input = input
+    @target_audio = nil
+
+    data = OPERATIONS[:video_input]
+    data += input.value.to_s(16).upcase.rjust(4, '0')
+
+    do_send("set parameter", data, name: "input", delay: 6.seconds)
+    video_input
+
+    # Double check the input again!
```

Comment:
Is there a need for this? Command will auto-retry if unsuccessful.

Reply:
I've got no idea but it's from the old ruby code so thought it might be needed. I've removed it for now and we can add it back in later if there are issues. https://github.com/PlaceOS/drivers/pull/50/commits/bad36ef655678893d95b21b2325362326df8980c